runsc/boot/network.go

```diff
@@ -184,6 +197,23 @@ func (r *Route) toTcpipRoute(id tcpip.NICID) (tcpip.Route, error) {
 	}, nil
 }
 
+func (n *Network) InitExternalStack(args *InitExternalStackArgs, _ *struct{}) error {
+	if stack := n.ExternalStack; stack != nil {
+		if err := stack.PreInitExternalStack(&externalstack.PreInitExternalStackArgs{}); err != nil {
```

Comment:
Why is initialization separated into 3 steps? Are they ever called in a different order or with other operations between steps?

Reply:
Taking TLDK for example, before actually initializing TLDK stack, we have some mount and namespace join to do. Also after initializing TLDK stack, there is some chroot to handle with the external stack. 

We are still working on how to refactor those operations into one CGO function call. Here just in case PreInitExternalStack and PostInitExternalStack will be invoked somewhere else, for now we separate initialization into three steps so they can be used separately. But if we can do initialization in one CGO function call, I think yes, we can just merge these three steps into one function.

Reply:
Yeah, if they're _always_ called like one-after-the-other like this, then I think it's best to have just one function. We can split it out if necessary, but it seems like it's not.

Reply:
Actually we are facing this problem during the refactoring of the TLDK stack setup, which may be similar to other network stacks as well. We need to prepare the necessary file descriptors (fds) required by TLDK, such as raw socket fds and vhost-net fds in virtio-user mode. Since these fds are specific to the network stack, we would like to invoke a CGO call to allow the network stack itself to handle this preparation.
Another aspect we need to consider is that this preparation needs to take place during the `runsc start` phase. If we try to perform these tasks within the `Init` function, which is invoked in the sandbox, we encounter issues due to namespacing and insufficient capabilities.
Therefore, the `PreInit` interface is needed. By utilizing this interface, we can prepare the fds before entering sandbox's namespace. Furthermore, if any mounts are set up during the `Init` phase, a corresponding `PostInit` interface is also needed to perform necessary cleanup operations after returning from netstack ns to sandbox ns.